### PR TITLE
Add dense hamiltonian

### PR DESCRIPTION
**Pull request by fdmalone**
Add a dense implementation of  factorised ERI (3-index) tensor for real hamiltonians. Avoids O(N^2M^2) memory overhead of storing half-rotated (ik|jl) used in the sparse code at a slightly increased computational cost. Ideal for larger molecules where sparsity in ERI tensor in MO basis is typically small. Works on GPU.

### CODE REVIEW

File: src/AFQMC/HamiltonianOperations/Real3IndexFactorization_batched_v2.hpp [hunk 1]
```diff
@@ -0,0 +1,629 @@
+//////////////////////////////////////////////////////////////////////
```
Comment (fdmalone) [inline src/AFQMC/HamiltonianOperations/Real3IndexFactorization_batched_v2.hpp]: Why is there _v2 version for these files. Multiple versions? Can we just include one?

File: external_codes/mpi_wrapper/mpi3/shared_window.hpp [hunk 2]
```diff
@@ -20,9 +20,9 @@ namespace mpi3{
 
 template<class T>
 struct shared_window : window<T>{
-//	shared_communicator& comm_;
-	shared_window(shared_communicator& comm, mpi3::size_t n, int disp_unit = alignof(T)) //: //sizeof(T)) : // here we assume that disp_unit is used for align
-		//window<T>()//, comm_{comm}
+	shared_communicator& comm_;
```
Comment (fdmalone) [inline external_codes/mpi_wrapper/mpi3/shared_window.hpp]: Please check merge conflict resolution.

File: src/AFQMC/Hamiltonians/RealDenseHamiltonian.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
+#include<cstdlib>
```
Comment (fdmalone) [inline src/AFQMC/Hamiltonians/RealDenseHamiltonian.cpp]: Has the bug for distributed Hamiltonians been fixed?

File: src/AFQMC/HamiltonianOperations/Real3IndexFactorization_batched_v2.hpp (hunk 1 above)
Comment (mmorale3) [reply src/AFQMC/HamiltonianOperations/Real3IndexFactorization_batched_v2.hpp]: Possibly. The original version is faster, but uses about 30% more memory. The v2 version is probably the one to keep, due to memory limitations on gpu. With c++17, we can keep both versions on a single file.

File: src/AFQMC/Hamiltonians/RealDenseHamiltonian.cpp (hunk 3 above)
Comment (mmorale3) [reply src/AFQMC/Hamiltonians/RealDenseHamiltonian.cpp]: not yet

**Comment by ye-luo** [comment]
Okay to test

**Pull request merged**